brscan5: init at 1.2.6-0

mattchrist opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
The existing brscan4 and dsseries packages and modules don't work correctly with my Brother DS-640. (with brscan4, scans would 'cut off' after 1/4 of the page.)
This PR creates a brscan5 driver package and NixOS 'hardware.sane.brscan5' module to set it up. 

This works for me with my scanner plugged in via USB, but I don't have a network scanner available to test with. If anyone has a compatible Brother network scanner, please try it out with this driver and report how it works.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

mattchrist commented:
@jraygauthier I'd appreciate you taking a look at this PR if you get a chance. Thanks!

jraygauthier commented:
Hi @mattchrist, thanks for contributing.

I might be wrong but at first glance it would seem to me there is a high incentive to extract some commonalities between both 4 and 5 and reuse these from both sides. Was there a lot a changes from the 4 code or were these minor?

mattchrist commented:
You're right, theres definitely an opportunity to refactor. I'll see what I can do. Thanks!

SuperSandro2000 commented on pkgs/applications/graphics/sane/backends/brscan5/preload.c:
```diff
@@ -0,0 +1,182 @@
+
+FILE* fopen(const char* path, const char* mode)
+{
+  char buf[PATH_MAX];
+	FILE* (*_fopen) (const char*, const char*) = dlsym(RTLD_NEXT, "fopen");
+
+	return _fopen(rewrite(path, buf), mode);
+}
+
```
Please do not mix tabs and spaces.

mattchrist replied:
Re-indented file.

SuperSandro2000 commented on pkgs/applications/graphics/sane/backends/brscan5/default.nix:
```diff
@@ -0,0 +1,106 @@
+
+  meta = {
+    description = "Brother brscan5 sane backend driver";
+    homepage = "http://www.brother.com";
```
```suggestion
    homepage = "https://www.brother.com";
```

SuperSandro2000 commented on pkgs/applications/graphics/sane/backends/brscan5/default.nix:
```diff
@@ -0,0 +1,106 @@
+    runHook postInstall
+  '';
+
+  dontStrip = true;
```
Why are we not striping?

mattchrist replied:
The 'brscan4' package which this is based on is not using stripping. I'll try it out with stripping.

mattchrist replied:
Not sure, but it works fine with stripping enabled. Fixed.

SuperSandro2000 commented on nixos/modules/services/hardware/sane_extra_backends/brscan5_etc_files.nix:
```diff
@@ -0,0 +1,69 @@
+
+  meta = with lib; {
+    description = "Brother brscan5 sane backend driver etc files";
+    homepage = "http://www.brother.com";
```
```suggestion
    homepage = "https://www.brother.com";
```

SuperSandro2000 commented on nixos/modules/services/hardware/sane_extra_backends/brscan5_etc_files.nix:
```diff
@@ -0,0 +1,69 @@
+    ${addAllNetDev netDevices}
+  '';
+
+  installPhase = ":";
```
```suggestion
  dontInstall = true;
```

SuperSandro2000 commented on nixos/modules/services/hardware/sane_extra_backends/brscan5_etc_files.nix:
```diff
@@ -0,0 +1,69 @@
+  addAllNetDev = xs: lib.concatStringsSep "\n" (map addNetDev xs);
+in
+
+stdenv.mkDerivation {
+
+  name = "brscan5-etc-files-1.2.6-0";
```
```suggestion
stdenv.mkDerivation {
  pname = "brscan5-etc-files";
  version = "1.2.6-0";
```

SuperSandro2000 commented on nixos/modules/services/hardware/sane_extra_backends/brscan5_etc_files.nix:
```diff
@@ -0,0 +1,69 @@
+
+  installPhase = ":";
+
+  dontStrip = true;
```
Why are we not striping?

SuperSandro2000 commented on pkgs/applications/graphics/sane/backends/brscan5/default.nix:
```diff
@@ -0,0 +1,106 @@
+    ${myPatchElf "opt/brother/scanner/brscan5/brscan_cnetconfig"}
+    ${myPatchElf "opt/brother/scanner/brscan5/brscan_gnetconfig"}
+
+    RPATH=${libusb1.out}/lib:${avahi-compat.out}/lib:${stdenv.cc.cc.lib}/lib:${gtk2}/lib:${glib.out}/lib
```
Please use makeLibPath instead.

SuperSandro2000 commented on pkgs/applications/graphics/sane/backends/brscan5/default.nix:
```diff
@@ -0,0 +1,106 @@
+  src = {
+    "i686-linux" = fetchurl {
+      url = "http://download.brother.com/welcome/dlf104033/${pname}-${version}.i386.deb";
+      sha256 = lib.fakeHash;
```
Missing hash.

mattchrist commented:
Marked PR as Draft. There's a few more things I want to do with this.

mattchrist committed:
brscan5: init at 1.2.6-0

mattchrist commented:
Result of `nixpkgs-review pr 123426` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>9 packages failed to build:</summary>
  <ul>
    <li>citrix_workspace</li>
    <li>citrix_workspace_20_04_0</li>
    <li>citrix_workspace_20_06_0</li>
    <li>citrix_workspace_20_09_0</li>
    <li>citrix_workspace_20_10_0</li>
    <li>citrix_workspace_20_12_0</li>
    <li>citrix_workspace_21_01_0</li>
    <li>gns3-server</li>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>37 packages built:</summary>
  <ul>
    <li>alerta-server</li>
    <li>appdaemon</li>
    <li>brscan5</li>
    <li>cura</li>
    <li>freeoffice</li>
    <li>gdbgui</li>
    <li>gns3-gui</li>
    <li>home-assistant</li>
    <li>libredirect</li>
    <li>moodle-dl</li>
    <li>octoprint</li>
    <li>onionshare</li>
    <li>onionshare-gui</li>
    <li>python38Packages.aioambient</li>
    <li>python38Packages.cassandra-driver</li>
    <li>python38Packages.flask-socketio</li>
    <li>python38Packages.python-engineio</li>
    <li>python38Packages.python-engineio_3</li>
    <li>python38Packages.python-socketio</li>
    <li>python38Packages.python-socketio_4</li>
    <li>python38Packages.sentry-sdk</li>
    <li>python38Packages.simplisafe-python</li>
    <li>python39Packages.aioambient</li>
    <li>python39Packages.cassandra-driver</li>
    <li>python39Packages.flask-socketio</li>
    <li>python39Packages.python-engineio</li>
    <li>python39Packages.python-engineio_3</li>
    <li>python39Packages.python-socketio</li>
    <li>python39Packages.python-socketio_4</li>
    <li>python39Packages.sentry-sdk</li>
    <li>python39Packages.simplisafe-python</li>
    <li>softmaker-office</li>
    <li>sublime-merge</li>
    <li>sublime-merge-dev</li>
    <li>sublime3</li>
    <li>sublime3-dev</li>
    <li>teamspeak_client</li>
  </ul>
</details>

mattchrist commented:
> * citrix_workspace
> 
> * citrix_workspace_20_04_0
> 
> * citrix_workspace_20_06_0
> 
> * citrix_workspace_20_09_0
> 
> * citrix_workspace_20_10_0
> 
> * citrix_workspace_20_12_0
> 
> * citrix_workspace_21_01_0
> 
> * gns3-server

These packages fail to build in master too.
- citrix_workspace* need manual downloads after accepting a eula (which I'm not interested in doing).
- gns3-server is missing a dependency 

nixos-install-tools I'm a bit confused about, it builds fine in my branch and in the nixpkgs-review shell:
```
[nix-shell:~/.cache/nixpkgs-review/pr-123426-2/nixpkgs]$ nix-build -A nixos-install-tools
/nix/store/kccc3lz58nanfby3qi5rb95ksi1irfsn-nixos-install-tools-21.05pre-git
```

mattchrist commented:
@SuperSandro2000 @jraygauthier: I'd appreciate you taking another look when you have a chance. Thanks!

jraygauthier commented on pkgs/build-support/libredirect/libredirect.c:
```diff
@@ -189,9 +190,85 @@ int posix_spawnp(pid_t * pid, const char * file,
     char buf[PATH_MAX];
     return execv_real(rewrite(path, buf), argv);
 }
+
```
Nice. We will be able to reuse this for brscan4 as well. I will keep that in mind. Might have implications on a couple other packages.

jraygauthier reviewed (approved):
LGTM. Better the the original.

mattchrist committed:
fix brscan5 config generation

before this, the config utility was unable to locate the models folder
update tests to use a compatible model

mattchrist commented:
Thanks for reviewing @jraygauthier and @mkg20001!

I just noticed that the brscan5 config utility wasn't properly 'identifying' network scanners. Pushed a small change to address that issue, and updated the NixOS test to catch that problem.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/383

jraygauthier reviewed (approved):
This last commit seems fine as well. I might attempt to run the thing on my computater and see if I am able to at least run the tests and attempt to register my scanner with brscan5, see what happens.

jraygauthier commented:
I took a second look at `libredirect.c` and was able to follow along new additions to the module.  I had to squint a bit for `replace_string` but finally understood the reason behind the second looping level `matches[nn] += len_diff` (an optimization to use a single buffer).  

```bash
$ nix-build -I nixpkgs=. -A libredirect
# Success here.
```

```bash
$ nix-build -I nixpkgs=. '<nixpkgs/nixos/tests/brscan5.nix>'
# Success here.
```

jraygauthier commented:
Re-basing on top of my current-config, using the following snippet:

```nix
{ config, pkgs, ... }:

{
  hardware.sane = {
    enable = true;
    # brscan4.enable = true;

    # brscan4.netDevices = {
    #   livingRoom = { model = "MFC-7860DW"; nodename = "BRW0080927AFBCE"; };
    # };

    brscan5.enable = true;
    brscan5.netDevices = {
      livingRoom = { model = "MFC-7860DW"; nodename = "BRW0080927AFBCE"; };
    };
  };
}
```

I was able to observe the following file system effect:

```bash
$ tree -l /etc/opt/brother/scanner/
/etc/opt/brother/scanner/
├── brscan5 -> /etc/static/opt/brother/scanner/brscan5
│   ├── brsanenetdevice.cfg
│   ├── brscan5.ini
│   └── models
│       ├── brscan5ext_0.ini
│       ├── brscan5ext_1.ini
│       ├── brscan5ext_2.ini
│       ├── brscan5ext_3.ini
│       ├── brscan5ext_4.ini
│       ├── brscan5ext_5.ini
│       ├── brscan5ext_6.ini
│       └── brscan5ext_7.ini
└── models -> /etc/static/opt/brother/scanner/models  [recursive, not followed]

3 directories, 10 files
```

```bash
$ cat /etc/opt/brother/scanner/brscan5/brscan5.ini 
[Support Model]

[ModelTypeName]

[Driver]
Browse=0
SKEYAPP=/opt/brother/scanner/brscan-skey/skey-scanimage

$ cat /etc/opt/brother/scanner/brscan5/brsanenetdevice.cfg 
DEVICE=livingRoom , "MFC-7860DW" , Unknown , NODENAME=BRW0080927AFBCE
```

Which still seems to make sense according to my spec. I would assume `Unknown` to mean that the configuration tool did not find the model I specified. Searching through `/etc/opt/brother/scanner/brscan5/models/*.ini` for my model as expected, I do not find it (as my model is expected to be supported by brscan4, not brscan5). 

Running scanimage, I obtain this:

```bash
$ scanimage -L
Segmentation fault (core dumped)
```

I am not sure if we should worry about the segmentation fault or not. Might have other cause (TBD).

jraygauthier commented on nixos/modules/services/hardware/sane_extra_backends/brscan5.nix:
```diff
@@ -0,0 +1,115 @@
+          The model of the network device.
+        '';
+
+        example = literalExample "MFC-7860DW";
```
Might be a good idea to change the example to a compatible brscan5 model.

mattchrist replied:
Nice catch, thanks! Pushed a fix.

jraygauthier commented:
Now, changing my spec for a brscan5 supported model (`model = "ADS-1200"`), as expected, I observed the following change:

```bash
$ cat /etc/opt/brother/scanner/brscan5/brsanenetdevice.cfg 
DEVICE=livingRoom , "ADS-1200" , 0x4f9:0x459 , NODENAME=BRW0080927AFBCE
```

Note how the `Unknown` cell was changed to `0x4f9:0x459 `.

I still observe the `scanimage` segfault however:

```bash
$ scanimage -L
Segmentation fault (core dumped)
```

Will have to investigate this. Maybe I need to restart?

jraygauthier commented:
Coming back from reboot.

With the unsupported model:

```bash
$ scanimage -L
# ..
device `brother5:net1;dev0' is a Brother
```

With the supported one:

```bash
$ scanimage -L
# ..
device `brother5:net1;dev0' is a Brother livingRoom ADS-1200
```

That's nice, as exactly as expected.

mattchrist committed:
brscan5: update example to be supported model

jraygauthier commented:
With the unsupported model spec (`"MFC-7860DW"` which is the actual model for my scanner HW), I was able to launch a scan of the flatbed and surprisingly still get a nice output using my usual scan soft.

Even more surprisingly likewise with support mode spec ("ADS-1200" that does not correspond to my HW). I would assume it fallback to a generic default scanner mode and one has to delve deeper into the model specific features in order to see a difference.

To me, this works perfectly. It now has been shown to work with another brother network scanner.

jraygauthier commented:
Really nice job @mattchrist.

mattchrist commented:
@Mic92 @aanderse Would you please review this PR?

aanderse commented on nixos/modules/services/hardware/sane_extra_backends/brscan5.nix:
```diff
@@ -0,0 +1,115 @@
+  options = {
+
+    hardware.sane.brscan5.enable =
+      mkEnableOption "Brother's brscan5 scan backend" // {
```
Does this actually render the way one might think it would? :thinking:

mattchrist replied:
`mkEnableOption "thing" // {description = ...` is done a few other modules (notably, the other sane_extra_backends):
```
nixos/modules/programs/java.nix
nixos/modules/programs/x2goserver.nix
nixos/modules/services/hardware/sane_extra_backends/brscan4.nix
nixos/modules/services/hardware/sane_extra_backends/dsseries.nix
nixos/modules/services/misc/gitea.nix
nixos/modules/services/networking/i2pd.nix
nixos/modules/services/networking/wpa_supplicant.nix
nixos/modules/services/web-apps/wiki-js.nix
nixos/modules/services/x11/xautolock.nix
nixos/modules/virtualisation/virtualbox-host.nix
```

Since this seems unusual I added a commit to simplify this option.

Thanks!

aanderse commented on nixos/modules/services/hardware/sane_extra_backends/brscan5.nix:
```diff
@@ -0,0 +1,115 @@
+    };
+  };
+
+  config = mkIf (config.hardware.sane.enable && cfg.enable) {
```
Can we drop the `.enable` option and check if `cfg.netDevices != {}` instead?

mattchrist replied:
`.enable` covers installation without netDevices to make USB scanners work. I think it needs to stay.

aanderse replied:
Thanks for confirming!

aanderse reviewed (commented):
I took a *really high level* look at the module. I don't have ~motivation~hardware to test this at the moment, but the overall approach of the module looks good :+1:

Let me know if the points I mentioned aren't valid, but otherwise as long as you are happy with the testing this module LGTM.

mattchrist committed:
brscan5: simplify mkEnableOption

mattchrist commented:
This PR has been reviewed and approved by multiple people, is there anything left that this request needs before it can be merged?
